### PR TITLE
Added embedding average

### PR DESCRIPTION
**Pull request by FelixLabelle**
If we are loading fast text, the target vectors which are present in the dataset but not the vocab are averaged and negated

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -28,7 +28,7 @@ def parse_args(args):
     # Modeling.
     parser.add_argument('--device', default='cpu', help='Select cuda for the gpu.')
     parser.add_argument('--model_name', default='lstm_attn')
-    parser.add_argument('--loss_function', default='xent')
+    parser.add_argument('--loss_function', default='xent', choices=['xent','vmf'])
```
Comment (artidoro) [inline main.py:31]: Nice touch, we should do this for the other options (like mode). Not in this pr though :)

File: utils.py [hunk 2]
```diff
@@ -74,12 +75,20 @@ def torchtext_iterators(
...
+            vectors=embedding_vectors
         )
         # TODO Remove words without embedding from the output vocabulary
         trg_field.vocab.vectors[trg_field.vocab.stoi[BOS_TOKEN]] = 1/16
         trg_field.vocab.vectors[(trg_field.vocab.vectors == 0).all(-1)] = trg_field.vocab.vectors.mean(0)
+        if is_vivo:
+            full_vocab = torchtext.data.Field()
+            full_vocab.build_vocab(train.trg, min_freq=min_freq)
+            excluded_words = [word for word in full_vocab.vocab.stoi if word not in trg_field.vocab.stoi]
+            excluded_vectors = torch.stack([embedding_vectors[excluded_word] for excluded_word in excluded_words if excluded_word in embedding_vectors.stoi])
```
Comment (artidoro) [inline utils.py]: Could you split long lines (just harder to look at on a smaller screen and github :) )

Comment (brendon-boldt) [inline utils.py]: If this list is empty, PyTorch complains and throws an error. This happens when I set `--max_size` to `40` or lower. It looks like `full_vocab` and `trg_field.vocab` are constructed in almost the same way which would mean that their set difference is `{}`.

**Review by brendon-boldt — CHANGES_REQUESTED** [review]

File: main.py (hunk 1 above)
Comment (FelixLabelle) [reply main.py:31]: Haha, I did that in the other PR (I was playing around there anyway).

File: utils.py (hunk 2 above)
Comment (FelixLabelle) [reply utils.py]: Yeah, will do!

Comment (FelixLabelle) [reply utils.py]: They are not though, full_vocab does not have a max size

Comment (FelixLabelle) [reply utils.py]: For the error I could add something, but I did not have that bug occur. How did you produce it?

Comment (brendon-boldt) [reply utils.py]:
```
python main.py \
    --device cuda \
    --checkpoint_path checkpoint_path \
    --fasttext_embeds_path corpus.fasttext.txt \
    --write_to_file \
    --eval_epochs 1 \
    --train_epochs 1 \
    --batch_size 64 \
    --max_len 40 \
    --loss_function vmf \
    --input_feed \
    --tie_embed \
    --enc_embed_size 300 \
    --mode train
```

Comment (brendon-boldt) [reply utils.py]: So maybe the error only occurs when your vocabulary is less than the max size.

Comment (FelixLabelle) [reply utils.py]: The vocab size for me is 112,000, which is twice the size of the max. I'll just check for an empty list in that case and return a random vector

Comment (brendon-boldt) [reply utils.py]: Even if you run with a `--max_len` of 40 or even 15?

Comment (FelixLabelle) [reply utils.py]: Ok I just pushed a verision that should fix your issue

Comment (FelixLabelle) [reply utils.py]: Yeah it works with short lengths now

Comment (FelixLabelle) [reply utils.py]: max len of 15,20,40 at the least

**Comment by brendon-boldt** [comment]
This worked for me now. Feel free to merge.

**Review by brendon-boldt — APPROVED** [review]

**Pull request merged**